### ISSUE TITLE
Hygiene issues with the use of syntax parameters as anaphoric variables

### ISSUE THREAD

**Issue by rocketnia** [racket/racket#3414]
**What version of Racket are you using?**

7.8 [cs].

First I tried some initial experiments on 7.7 [bc], and then I decided to upgrade in case this behavior had already changed.

**What program did you run?**

```racket

(define-syntax-parameter it (syntax-rules ()))

(define-syntax let-it
  (syntax-rules ()
    [(let-it val body)
     (let ([var val])
       (syntax-parameterize ([it (syntax-id-rules () [_ var])])
         body))]))

(let-it "outer"
  (let-syntax ([old-it (syntax-id-rules () [_ it])])
    (let-it "inner"
      old-it)))
```

The last expression returns `"inner"`.

**What should have happened?**

I believe it should have returned `"outer"`.

In most discussions of syntax parameters I can find, they're touted as a way to implement anaphoric variables. The documentation even includes an example of implementing `aif`, an anaphoric `if`. I took that example and simplified it to write the `let-it` utility above, just to make sure I wasn't doing anything too exotic.

It seems to me a correct implementation of anaphoric variables would act as a shorthand for using explicit variable bindings, like so:

```racket
(let ([x "outer"])
  (let-syntax ([old-x (syntax-id-rules () [_ x])])
    (let ([x "inner"])
      old-x)))
```

However, the code with explicit bindings returns `"outer"`, while the code with syntax-parameter-based anaphors returns `"inner"`.

Speculating a bit, I suspect syntax parameter bindings and lookups aren't currently being associated with sets of scopes. It seems like they _could_ be, and if they're not, that could simply be a bug.

If this is a bug, it seems fixable. Here's a demonstration of an alternative approach, making use of a gensym variable name to carry the set of scopes. Using this approach, the example returns `"outer"` as I would expect:

```racket
(define-for-syntax that-id (gensym "that"))

(define-syntax (that stx)
  (datum->syntax stx that-id))

(define-syntax (let-that stx)
  (syntax-case stx ()
    [(let-that val body)
      #`(let ([#,(datum->syntax stx that-id) val])
          body)]))

(let-that "outer"
  (let-syntax ([old-that (syntax-id-rules () [_ that])])
    (let-that "inner"
      old-that)))
```

Like the syntax parameter approach, this gensym approach is "hygienic" in the sense that its users are able to shadow the `that` binding and control what names (if any) it's imported under. Considering this scenario, I might say it's even more hygienic in the sense that the anaphoric variable's lexical binding is respected by syntax-quoting forms like `syntax-id-rules`.

I should note there's another situation for which these approaches differ in behavior:

```racket
; Syntax parameter approach

(define-syntax-rule (using-it-as-an-implementation-detail body)
  (let-it "inner"
    body))

; Returns "inner"
(let-it "outer"
  (using-it-as-an-implementation-detail
    it))


; Explicit variable bindings

(define-syntax-rule (using-x-as-an-implementation-detail body)
  (let ([x "inner"])
    body))

; Returns "outer"
(let ([x "outer"])
  (using-x-as-an-implementation-detail
    x))


; Gensym approach

(define-syntax-rule (using-that-as-an-implementation-detail body)
  (let-that "inner"
    body))

; Returns "outer"
(let-that "outer"
  (using-that-as-an-implementation-detail
    that))
```

This once again looks good for the gensym approach, but I think there's some subtlety here that makes it harder to decide.

Specifically, someone might want to write an anaphoric binding macro that does its work by expanding into another anaphoric binding macro. From that person's perspective, `using-it-as-an-implementation-detail` has exactly the kind of implementation that should make it an anaphoric binder (regardless of what its name is).

I think this wiggle room stems from ambiguity in the very idea of generating a variable that's visible to "the user." Should the variable be visible to the immediate user of `let-it` or to the indirect user of `using-it-as-an-implementation-detail`? In other words, are we trying to make the `it` binding break through one macro's hygiene, or are we trying to designate it as a permanently special binding that breaks through every macro call boundary it comes across?

I'm inclined to prefer an option that breaks through fewer macro call boundaries. Racket macros are hygienic by default, and when someone writes `(let ([it ...]) ...)` in a macroexpansion result, they won't get an `it` binding that's visible to the user. They need to be more explicit to do that. If they find out `(let-it ...)` isn't explicit enough either, it should be just as easy to understand the rationale.

I realize changing the behavior of syntax parameters now would raise issues of backwards compatibility. I can imagine at least a few resolutions:

* Changing the behavior of syntax parameters to fix these scenarios, even in spite of the backwards compatibility cost.

* Keeping syntax parameters the way they are, but changing the documentation to recommend using a less extreme form of hygiene-breaking for simple anaphoric utilities like `aif`.

* Adding a keyword argument or alternative API people can use with syntax parameters to say they want the binding sites' sets of scopes to be taken into account. This kind of configuration might be supplied when creating, binding, and/or accessing a syntax parameter.

* Of course, just embracing it as it is. Maybe people really do intend this kind of behavior when they use anaphoric variables.

**Comment by jackfirth** [comment]
Huh, I assumed that this:

```scheme
(let-it "outer"
  (let-syntax ([old-it (syntax-id-rules () [_ it])])
    (let-it "inner"
      old-it)))
```

should "obviously" return `"inner"`, since that's how parameters would work. For example this code:

```scheme
(define it (make-parameter 'default))

(define (call/it new-it body)
  (parameterize ([it new-it]) (body)))

(call/it "outer"
  (lambda ()
    (let ([old-it it])
      (call/it "inner"
        (lambda ()
          (it))))))
```

evaluates to `"inner"`, which makes sense to me because `old-it` isn't referencing the _current value_ of `it`, it's referencing _the entire parameter_. A `let-syntax` binding with a syntax parameter works the same way.

**Comment by rocketnia** [comment]
> @rocketnia: [...] someone might want to write an anaphoric binding macro that does its work by expanding into another anaphoric binding macro [...]

Oh, this was more on the mark than I remembered.

I know I've read "[Keeping It Clean with Syntax Parameters](http://scheme2011.ucombinator.org/papers/Barzilay2011.pdf)" before, and I took another look at it. (This isn't anywhere near where syntax parameters started, right? It's just a nice exploration of the application of syntax parameters to the purpose of anaphors.) Even on the first page, it's obvious that part of the goal is to expose anaphoric bindings in ways that implicitly escape from multiple layers of macro abstraction:

> The problem with [individual uses of `syntax->datum`] is that it does not compose well with new macros that expand to uses of `forever` [an example macro that anaphorically binds `abort`]. For example, suppose that a `while` macro is defined as follows, with a goal of having `abort` as well:
>
> ```racket
> (define-syntax while
>   (syntax-rules ()
>     [(while test body ...)
>      (forever (unless test (abort)) body ...)]))
> ```
>
> [...]

The system demonstrated on pages 2 and 3 (`define-syntax-rules/capture`) approaches anaphors in a way where each macro has to explicitly list what anaphoric bindings it exposes. The paper remarks,

> This requirement is impractical: some macros in the chain might come from libraries that are not under our control, and composing macros with different unhygienic key-words makes for additional explicit linking. If we wish to create a language where a fundamental form like `if` is extended with such a keyword to create an anaphoric conditional, then we would need to link up the introduced `it` in _any_ derived macros. This makes the effort of constructing and maintaining such languages prohibitively expensive.

I'm not especially convinced. If someone has a long list of variable names they have to write over and over, they can define a macro to automate that. In this way, the `define-syntax-rules/capture` approach could continue to be built up in a way that keeps macros explicit about how they interact with variable scoping.

But at this point I'm uncovering objections I have to explicit goals of this technique, not uncovering a bug in the implementation. At any rate, the behavior I've reported isn't new or unheard of, and it's probably not a mistake now if it ever was. Meanwhile, it doesn't stop me from expanding upon my gensym example or `define-syntax-rules/capture` if I prefer those semantics.

This issue should probably be closed now. The most productive way to "resolve" this would probably be to update the documentation to avoid misconceptions like mine, but I don't have specific suggestions.

---

> @jackfirth: [...] since that's how parameters would work

Yeah, syntax parameters and (run time) parameters are very much alike as you say. I'm glad you illustrated that connection. (Keeping It Clean also goes into depth on this.)

When opening this issue, I thought of syntax parameters as merely an implementation technique in this situation.

By way of analogy, consider the bindings made by `let-syntax`. They pretty clearly have lexical scope, but during macroexpansion, we look them up using `syntax-local-value`, which makes use of the dynamically scoped expansion context. We as the user of `syntax-local-value` see dynamic scope, but we as the user of `let-syntax` see lexical scope. Dynamic scope is just part of the implementation.

For some time, I've thought `syntax-parameterize` might as well work using the gensym technique above, essentially not using any dynamic scope at any level. But it turns out they do have some pretty significant `parameterize`-like qualities, and I just hadn't run into the scenarios that proved it. :)
